### ISSUE TITLE
Remove the remaining uses of `old_orphan_rules`

### ISSUE THREAD

**Issue by arielb1** [rust-lang/rust#19470]
**UPDATED ISSUE:**

The original issue that was reported by @arielb1 has largely been resolved. However, there remain some traits that were added before the new rules were completed, and which do not conform. Those traits are currently distinguished with `old_orphan_check` markers. The remaining work is to fix these traits:

- [x] `BorrowFrom`
- [x] `BorrowFromMut`
- [x] `IntoCow`
- [x] `ToOwned`
- [ ] `PartialEq`

\- @nikomatsakis

**ORIGINAL ISSUE:**

The current version of orphan checking does not prevent me from break coherence if impls from different crates are non-orphan because of different type parameters of the same type, for example:

base.rs:
```
pub trait Base { fn work(&self); }
```

foo.rs:
```
extern crate base;
pub struct Foo;

impl<T> base::Base for (Foo, T) {
  fn work(&self) { println!("foo!"); }
}
```

bar.rs:
```
extern crate base;
pub struct Bar;

impl<T> base::Base for (T, Bar) {
  fn work(&self) { println!("bar!"); }
}
```

foobar.rs:
```
extern crate foo;
extern crate bar;
extern crate base;
use base::Base;

fn main() { (foo::Foo, bar::Bar).work(); }
```

Which of course ICE-s with a trans ambiguity.

**Comment by arielb1** [comment]
Looking at this again, I *do* want to be able to do `impl<D:Decoder<E>, E> Decodable<D, E> for InternedString`.

**Comment by aturon** [comment]
@nikomatsakis 

> Another interesting example that fails (from libgraphviz):
> 
> ```rust
> impl<'a, T: PartialEq, Sized? V: AsSlice<T>> Equiv<V> for MaybeOwnedVector<'a, T>
> ```
> 
> This seems to be a pattern that is just inexpressible?
> 
> (That is, relating a local type to *any* type that implements some trait).

Somehow I did not fully think through the implications of this limitation before. I think it's a serious one and I agree with @arielb1 that we should consider a less conservative form of the rules that can allow it.

**Comment by nikomatsakis** [comment]
So @aturon and I had a long conversation about this and we've decided that @arielb1's original ordering-based solution is probably the best we are going to do. It is somewhat unfortunate, but there are critical patterns that are just not supported by the coverage-based rules. We tried to find a way to combine the two rules, but the most important pattern seems inherently unsupportable:

    impl<T> BorrowFrom<Rc<T>> for T

As long as you permit something like:

    impl<T> BorrowFrom<T> for LocalType

then there will be an inherent conflict. Luckily we are planning to change `BorrowFrom` to `Borrow`, so maybe this just becomes a thing -- you must order your type parameters well.

Similar problems arise with binops, but there the problems run deeper:

    impl<E,T:Iterator<Item=E>> Add<T> for MyVec<E>

This pattern is ok, but changing the order of the add operands is not. However, that seems to just be...what it is. And what's more, if we DID permit both orders, there's an obvious conflict between any two vector implementations that want to be addable to all iterators. So forcing an ordering makes those semantics clear.

Perhaps we'll find a way to improve this rule in the future; it should be backwards compatible to accept more impls.

Sorry @arielb1 for taking so long to come around!

**Comment by nikomatsakis** [comment]
So I implemented the partial ordering rule. Here are the traits that fail, and a sample impl from each one:

- `BorrowFrom`
- `BorrowFromMut`
- `IntoCow` -- `impl<'a, T> IntoCow<'a, Vec<T>, [T]> for &'a [T]`
- `ToOwned` -- `impl<T: Clone> ToOwned<Vec<T>> for [T]`
- `PartialEq` -- `impl<'b, A, B> PartialEq<Vec<A>> for &'b [B]`

**Comment by nikomatsakis** [comment]
(While it's not clear what the best sol'n is for each of these traits, it seems like they are all solvable.)

**Comment by nikomatsakis** [comment]
So out of curiosity, I also implemented a more restrictive rule:

1. The self type must reach something local
2. Any type parameters in the self type must be "covered by" something local

This also lets everything in the standard library build. It does cause failures on some test cases that the other check accepted, such as:

- `impl Remote<BigInt> for int`
- `impl Remote<BigInt> for Vec<int>`

Both of these are interesting patterns, but otoh they can't be scaled up to blanket impls over all `T:ToInt` or anything like that.

The advantage of the simpler check is primarily that it's easier to explain. It also doesn't commit us to the full ordering rules just yet, in case we come up with something better (though I'm not so worried about that). The disadvantage is that we lose some impls.

Thoughts?

**Comment by nikomatsakis** [comment]
Ah, I was wrong, there one additional failure that I didn't notice before using the more restrictive rules, and it's a semi-interesting one:

- `impl<'a, T> FromIterator<T> for Cow<'a, Vec<T>, [T]>`

Here the type parameter `T` appears uncovered, even though it also appears covered.

This gives me an idea for an alternative rule. One could imagine:

1. The self type must reach something local
2. All type parameters in the self type must be covered by a local type in the self type

This allows type parameters to appear in arbitrary other places. I believe this is sound, though, or at least as sound as the original typing rule -- it's basically the covering rule applied only to the self type. 

One reason this rule is interesting is that it does permit for truly order-independent traits, though they must be implemented using tuples in the self type. (Rather like my original max-min proposal for multidispatch.) Not saying we will use this pattern, but it leaves the door open.

I implemented this rule too. It builds all stdlibs as well.

**Comment by aturon** [comment]
@nikomatsakis and I talked about this last proposal:

> 1. The self type must reach something local
> 2. All type parameters in the self type must be covered by a local type in the self type

and think it seems very promising. On the one hand, it covers the most important impl patterns we want to cover. On the other hand, it's relatively easy to think about: to a rough approximation, only the definer of a trait can give a blanket impl for `Self`; downstream crates must impl "on" a type defined in their crate. (Of course the full rule is more subtle, but that gets you most of the way there.)

It also means that the order of type parameters does not become significant for `impl` purposes *other* than `Self` -- and `Self` is already special in various other ways.

Finally, it seems plausible to extend this rule in the future, allowing traits to mark additional non-`Self` parameters as participating in the same locality checks.

**Comment by glaebhoerl** [comment]
Interestingly this seems very similar to a bug [GHC also has](http://www.reddit.com/r/haskell/comments/2agy14/type_classes_confluence_coherence_global/cjynv30?context=3).

**Comment by aturon** [comment]
Nominating to move to the -beta milestone.

**Comment by nikomatsakis** [comment]
I updated the issue to better highlight what remains to be done, and made a checklist of traits that are still using `old_orphan_check` for one reason or another.

**Comment by aturon** [comment]
Nominating to move to P-backcompat-libs.

This will largely go away with the stabilization of `std::borrow` (which is currently blocked) but there are a couple other instances lurking.

**Comment by pnkfelix** [comment]
P-backcompat-libs, 1.0 beta.
